### ISSUE TITLE
When we called the 'get' method in a modified field on PreUpdate event, it loses value

### ISSUE THREAD

**Issue by meigallodixital** [doctrine/mongodb-odm#1070]
Preupdate code:

```
   public function preUpdate(PreUpdateEventArgs $eventArgs)
    {
        $documento = $eventArgs->getDocument();

        if (!$documento instanceof Producto) {
            return;
        }

        $dm = $eventArgs->getDocumentManager();
        $uow = $dm->getUnitOfWork();
       
        
        // Actualizamos importes
        if (!is_numeric($documento->getEntrada()->getImportes()->getDescuento())) {
            $documento->getEntrada()->getImportes()->setDescuento(0);
        }
        $documento->getEntrada()->getImportes()->setEuros($documento->getEntrada()->getImportes()->getComprado() - $documento->getEntrada()->getImportes()->getDescuento());

        // Regeneramos el documento
        $meta = $dm->getClassMetadata(get_class($documento));
        $uow->recomputeSingleDocumentChangeSet($meta, $documento);

    }
```

If we change 'comprado' field value lost new value, but if we remove
```
$documento->getEntrada()->getImportes()->getComprado()
```
New value is saved.

**Comment by malarzm** [comment]
@meigallodixital could you transform this into failing test case?

**Comment by meigallodixital** [comment]
**Test class:**
```
<?php

namespace TestProject\ProductoBundle\Tests\Controller;

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;


class DefaultControllerTest extends KernelTestCase
{
	/**
	 * @var \Doctrine\ODM\DocumentManager
	 */
	private $dm;

	/**
	* {@inheritDoc}
	*/
	public function setUp()
	{
		self::bootKernel();
		$this->dm = static::$kernel->getContainer()
			->get('doctrine_mongodb')
			->getManager();
	}

    public function testComprado()
    {
    	$producto = $this->dm->getRepository('ProductoBundle:Producto')->findOneByCodigo('14031700095');
        if (!$producto) {
            throw $this->createNotFoundException($this->get('translator')->trans('producto.no.encontrado'));
        }
        $producto->getEntrada()->getImportes()->setDescuento(0);
        $producto->getEntrada()->getImportes()->setComprado(20);
        $this->dm->flush();
        $this->assertEquals(20, $producto->getEntrada()->getImportes()->getComprado());
        $producto2 = $this->dm->getRepository('ProductoBundle:Producto')->findOneByCodigo('14031700095');
        $this->assertEquals(20, $producto2->getEntrada()->getImportes()->getComprado());
    }
}
```
**'comprado' field initial value:** 10.
**Asserts:** OK (1 test, 2 assertions)

With getComprado() method active on preUpdate:
```
{
    "entrada" : {
        "importes" : {
            "descuento" : 0,
            "comprado" : 10,
            "euros" : 20
        }
    }
}
```

With getComprado() method removed on preUpdate:
```
{
    "entrada" : {
        "importes" : {
            "descuento" : 0,
            "comprado" : 20,
            "euros" : 20
        }
    }
}
```

If we print comprado value just before  of  ```$uow->recomputeSingleDocumentChangeSet($meta, $documento);``` command, it gives 20 in two cases.

**Comment by malarzm** [comment]
I was thinking about failing unit test that I could actually run by myself, I still don't know your mapping or what's really happening in your documents :)

**Comment by meigallodixital** [comment]
I made a *stupid micro app* with a simple form with the event and have same behaviour.

[Link to app] (https://drive.google.com/file/d/0B8vL4J3n5FV4OWpCVDRRbEVreUk/view?usp=sharing)

This is an application with a simple form (without validations, etc) + the event. Field **codigo** must be set manually ;)

In the first submit with *comprado* changed field, the event does subtraction correctly, but does not save the new value of *comprado*, continues saving old value. If we do not change the value of *comprado* and resend this is stored properly, but if we again change value continues without saving.

In other words:

if field **comprado** have value of 100, send 200 and next send 250 continues saving 100.
if field **comprado** have value of 100, send 250 and next send 250 it saves first time 100 and second 250.

In all cases substraction is done correctly with the new value.

I really appreciate any help you can provide.

**Comment by malarzm** [comment]
@meigallodixital I've run the application you sent and managed to execute it after getting rid of PHP error in Controller but I fail to see what is/should be happening over there. It would make things a lot easier to have (failing) test case prepared like these [here](https://github.com/doctrine/mongodb-odm/tree/master/tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket) as we can use it in future as regression test. Also can you please use English while writing code? It is far more readable then :)

**Comment by meigallodixital** [comment]
There are a missing $ in the controller on example app :(

We think that we found the problem an hour ago.

We were thinking that a document PreUpdate mongodb event with embebed subdocuments was unique. In the tests we have seen that the event listener makes a Preupdate for each class doctrine document, in this case 3 calls. We are not sure who was doing recomputeSingleDocumentChangeSet method with the previous configuration. 

We're now changing events for each embedded class document. We think that this was the origin of the problem.

My question now is how  we can access from a event of a class embedded  to parent  document data.

**Comment by malarzm** [comment]
[UnitOfWork::getParentAssociations](https://github.com/doctrine/mongodb-odm/blob/683728eb0b4723dd5551686addc724f8f1b13afe/lib/Doctrine/ODM/MongoDB/UnitOfWork.php#L320) should help you if I remember correctly :)

**Comment by meigallodixital** [comment]
thanks a lot!

**Comment by malarzm** [comment]
@meigallodixital closing the issue, feel free to reopen if anything would go wrong on ODM's side :)
